src/test/unit/link_quality_unittest.cc

```diff
@@ -0,0 +1,534 @@
+/*
+ * This file is part of Cleanflight.
+ *
+ * Cleanflight is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Cleanflight is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Cleanflight.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <string.h>
+
+extern "C" {
+    #include "platform.h"
+    #include "build/debug.h"
+
+    #include "blackbox/blackbox.h"
+    #include "blackbox/blackbox_io.h"
+
+    #include "common/time.h"
+
+    #include "drivers/max7456_symbols.h"
+    #include "drivers/persistent.h"
+    #include "drivers/serial.h"
+    #include "drivers/io.h"
+
+    #include "fc/config.h"
+    #include "fc/core.h"
+    #include "fc/rc_controls.h"
+    #include "fc/rc_modes.h"
+    #include "fc/runtime_config.h"
+
+    #include "flight/mixer.h"
+    #include "flight/pid.h"
+    #include "flight/imu.h"
+
+    #include "io/beeper.h"
+    #include "io/gps.h"
+
+    #include "osd/osd.h"
+    #include "osd/osd_elements.h"
+
+    #include "pg/pg.h"
+    #include "pg/pg_ids.h"
+    #include "pg/rx.h"
+
+    #include "rx/rx.h"
+
+    #include "sensors/battery.h"
+
+    attitudeEulerAngles_t attitude;
+    pidProfile_t *currentPidProfile;
+    int16_t rcData[MAX_SUPPORTED_RC_CHANNEL_COUNT];
+    uint8_t GPS_numSat;
+    uint16_t GPS_distanceToHome;
+    int16_t GPS_directionToHome;
+    uint32_t GPS_distanceFlownInCm;
+    int32_t GPS_coord[2];
+    gpsSolutionData_t gpsSol;
+    float motor[8];
+    float motorOutputHigh = 2047;
+    float motorOutputLow = 1000;
+    acc_t acc;
+    float accAverage[XYZ_AXIS_COUNT];
+
+    PG_REGISTER(batteryConfig_t, batteryConfig, PG_BATTERY_CONFIG, 0);
+    PG_REGISTER(blackboxConfig_t, blackboxConfig, PG_BLACKBOX_CONFIG, 0);
+    PG_REGISTER(systemConfig_t, systemConfig, PG_SYSTEM_CONFIG, 0);
+    PG_REGISTER(pilotConfig_t, pilotConfig, PG_PILOT_CONFIG, 0);
+    PG_REGISTER(imuConfig_t, imuConfig, PG_IMU_CONFIG, 0);
+
+    timeUs_t simulationTime = 0;
+
+    void osdRefresh(timeUs_t currentTimeUs);
+    uint16_t updateLinkQualitySamples(uint16_t value);
+#define LINK_QUALITY_SAMPLE_COUNT 16
+}
+
+/* #define DEBUG_OSD */
+
+#include "unittest_macros.h"
+#include "unittest_displayport.h"
+#include "gtest/gtest.h"
+
+extern "C" {
+PG_REGISTER(flight3DConfig_t, flight3DConfig, PG_MOTOR_3D_CONFIG, 0);
```

Comment:
Indentation is wrong here.

Reply:
will clean up.
1 question 
in crsf.c i have  
#ifdef USE_RX_LINK_QUALITY_INFO
STATIC_UNIT_TESTED uint16_t scaleCrsfLq(uint16_t lqvalue) {

but its not actual called i could not link to it ...
and had to use a copy of the same method   in the unit test it this ok. ?

Reply:
You should not have to. And if you want to test the function, then you should obviously call it, and not a copy of it.
You might have to define `USE_RX_LINK_QUALITY_INFO` for `link_quality_unittest` in `src/test/Makefile`.

Reply:
mike this had me stump for days

USE_RX_LINK_QUALITY_INFO is defined .

the only way is to move the function to rx.c ? but that is ugly ..

when i added crsf.c crsf.h ect. i when down a rabbit hole. in linker issues.

it was so bad i almost drooped the unit test i made .
i used it for it for sanity checks of the code.  and it did find a few.

Reply:
Ah I see. What you need is a declaration of the function in your test, since there isn't one in the header file:

```
uint16_t scaleCrsfLq(uint16_t lqvalue);
```

Reply:
tinker with unit test time .. 
holding off until the confusion issue is resolved. :)

Reply:
cleaned up spaces 
add scaleCrsfLq to unit test